---
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)
library(UpSetR)
source("../../util/plotting_functions.R")
source("../../util/intersect_functions.R")
source("../../util/_setup.R")
```


```{r}

# test
# 
# KOiTg_48_IP_R1_peaks.broadPeak
# 
# Controls
# 
# KOiTg_IP_R1_peaks.broadPeak
# 
# KO_CTL_IP_R1_peaks.broadPeak
# 
# KO_CTL_48_IP_R1_peaks.broadPeak

```


```{r}
gencode_gr <- rtracklayer::import("/scratch/Shares/rinn/genomes/Mus_musculus/Gencode/M25/gencode.vM25.annotation.gtf")

fl <- list.files("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/broad_peaks", full.names=TRUE)
peaks <- lapply(fl, rtracklayer::import)
names(peaks) <- gsub("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/broad_peaks/|_peaks.broadPeak", "", fl)


all_peaks_gr <- c(peaks[[1]], peaks[[2]], peaks[[3]], peaks[[4]],
                  peaks[[5]], peaks[[6]], peaks[[7]], peaks[[8]])

all_peaks_gr <- GenomicRanges::reduce(all_peaks_gr)
all_peaks_gr$peak_id <- paste0("peak_", 1:length(all_peaks_gr))

# Construct peak overlap matrix
peak_ov <- matrix(numeric(), ncol = length(all_peaks_gr))
colnames(peak_ov) <- all_peaks_gr$peak_id
for(i in 1:length(peaks)) {
  ov <- countOverlaps(all_peaks_gr, peaks[[i]])
  ov <- ov > 0
  ov <- as.numeric(ov)
  peak_ov <- rbind(peak_ov, ov)
  rownames(peak_ov)[[i]] <- names(peaks)[[i]]
}

upset(peak_ov)

peak_ov_df <- peak_ov %>% t() %>% as.data.frame()
pdf("figures/ko_ctrl_upset_plot.pdf")
upset(peak_ov_df, nsets = 8, sets = c("KOiTg_IP_R1","KO_CTL_IP_R1", "KO_CTL_48_IP_R1"))
dev.off()

```

```{r}
ctl_group <- c("KOiTg_IP_R1","KO_CTL_IP_R1", "KO_CTL_48_IP_R1")
trt_group <- c("KOiTg_48_IP_R1")

peak_ov_df <- peak_ov %>% t() %>% as.data.frame() %>% rownames_to_column("peak_id") %>%
  mutate(ko_control_consensus = as.numeric(KOiTg_IP_R1 == 1 & KO_CTL_IP_R1 == 1 & KO_CTL_48_IP_R1 == 1))

ko_trt <- peak_ov_df %>%
  group_by(ko_control_consensus, KOiTg_48_IP_R1) %>%
  summarize(count = n())
ko_trt
hmm <- peak_ov_df %>%
  group_by(ko_control_consensus, WT_CTL_IP_R1) %>%
  summarize(count = n())
hmm
ko_trt <- peak_ov_df %>% dplyr::select(ko_control_consensus, KOiTg_48_IP_R1)

```









